####Name : Sourabh Tiwari
####Code : DST 20923



## Explanation :
This Database Contains Movies Created By Several Directors In The World.

This Database Can Be Seen At: https://www.kaggle.com/luizpaulodeoliveira/imdb-project-sql

This Database Contains Two Tables. The First Table Is Called 'directors', Contains The Name Of The Director And His Gender. The Second Table Is Called 'movies', Contains The Title Of The Movies, budget, revenue, popularity, etc.

From This database, You Can Find Out The Best Movies Based On Income, Popularity And Votes. In Addition, You Can Also Find Out The Best Director Based On The Amount Of Revenue From All His Movies.

The Correlation Coefficient Of These Two Tables Can Show The Effect Of Each Column To The Other Column.

--> Analysis on IMDB useful libraries with sql
### Explanation
__SQLite__ : - SQLite database is integrated with the application that accesses the database. The applications interact with the SQLite database read and write directly from the database files stored on disk.

SQLite does NOT require a server to run.

Refer at https://www.sqlite.org/index.html

__PANDAS__ : - Pandas provide high performance, fast, easy to use data structures and data analysis tools for manipulating numeric data and time series. Pandas is built on the numpy library and written in languages like Python, Cython, and C. In pandas, we can import data from various file formats like JSON, SQL, Microsoft Excel, etc.

__NUMPY__: - Numpy is an Python library,It is used for scientific computing in python.It contains a collection of tools and techniques that can be used to solve on computers with mathematical model of problems. - High performance and multi-dimensional areay objects. - High level mathematical functions. - matrices

In [ ]:
# Import Pandas and NumPy

import numpy as np
import pandas as pd

#pip install db-sqlite3
#SQLite in general is a server-less database that you can use within almost all programming languages including Python.
import sqlite3


We import the sqlite3 module and then create a connection object which will connect us to the database and will let us execute the SQL statements.

sqlite3.connect(Database)

__cursor()__ - we can use the cursor object to call the execute() method to execute any SQL queries.

__STEPS:__
To query data in an SQLite database from Python, you use these steps:

- First, establish a connection to the SQLite database by creating a Connection object.
- Next, create a Cursor object using the cursor method of the Connection object.
- Then, execute a SELECT statement.
- After that, call the fetchall() method of the cursor object to fetch the data.
- Finally, loop the cursor and process each row individually.

In [ ]:
db = 'movies.sqlite'


conn = sqlite3.connect(db)
cur = conn.cursor()


In [ ]:
cur.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cur.fetchall()
print(tables)


[('movies',), ('sqlite_sequence',), ('directors',)]


## QUESTIONS
### Question 1: Can you get all data about movies?

In [ ]:
cur.execute('SELECT * FROM movies')
#use the cursor object to call the execute() method to execute any SQL queries.
movies = cur.fetchall()

#movies -> if u print movies , all the data shown in list manner, can't understanadble

#Note - fetchall() --> fetch all rows for the current query, if no rows avaliable then,it returns an empty list.

movies = pd.DataFrame(movies, columns = ['id', 'original_title', 'budget', 'popularity', 'release_date', 'revenue', 'title', 'vote_average', 'vote_count', 'overview', 'tagline', 'uid', 'director_id'])
#Creating a dataframe, A dataframe contains rows and columns , It is a Two-dimensional, size-mutable, potentially heterogeneous tabular data.



print('Shape of Movies data :',movies.shape)
print('\n\n\n')
movies.head(3)
#Displays first 3 rows in the table


Shape of Movies data : (4773, 13)






id                            original_title     budget  popularity  \
0  43597                                    Avatar  237000000         150   
1  43598  Pirates of the Caribbean: At World's End  300000000         139   
2  43599                                   Spectre  245000000         107   

  release_date     revenue                                     title  \
0   2009-12-10  2787965087                                    Avatar   
1   2007-05-19   961000000  Pirates of the Caribbean: At World's End   
2   2015-10-26   880674609                                   Spectre   

   vote_average  vote_count  \
0           7.2       11800   
1           6.9        4500   
2           6.3        4466   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                          tagline     uid  director_id  
0                     Enter the World of Pandora.   19995         4762  
1  At the end of the world, the adventure begins.     285         4763  
2                           A Plan No One Escapes  206647         4764

In [ ]:
cur.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cur.fetchall()
print(tables)


[('movies',), ('sqlite_sequence',), ('directors',)]


In [ ]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4773 entries, 0 to 4772
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              4773 non-null   int64  
 1   original_title  4773 non-null   object 
 2   budget          4773 non-null   int64  
 3   popularity      4773 non-null   int64  
 4   release_date    4773 non-null   object 
 5   revenue         4773 non-null   int64  
 6   title           4773 non-null   object 
 7   vote_average    4773 non-null   float64
 8   vote_count      4773 non-null   int64  
 9   overview        4770 non-null   object 
 10  tagline         3951 non-null   object 
 11  uid             4773 non-null   int64  
 12  director_id     4773 non-null   int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 484.9+ KB


## Question 2: How do you get all data about directors?

In [ ]:
cur.execute('SELECT * FROM directors')
#use the cursor object to call the execute() method to execute any SQL queries.

directors = cur.fetchall()
#Note - fetchall() --> fetch all rows for the current query, if no rows avaliable then,it returns an empty list.


directors = pd.DataFrame(directors, columns = ['name', 'id', 'gender', 'uid', 'department'])
#Creating a dataframe, A dataframe contains rows and columns , It is a Two-dimensional, size-mutable, potentially heterogeneous tabular data.

print('Shape of Directors data :',directors.shape)
print('\n\n\n')


directors.head(100)

Shape of Directors data : (2349, 5)






name    id  gender     uid department
0       James Cameron  4762       2    2710  Directing
1      Gore Verbinski  4763       2    1704  Directing
2          Sam Mendes  4764       2      39  Directing
3   Christopher Nolan  4765       2     525  Directing
4      Andrew Stanton  4766       2       7  Directing
..                ...   ...     ...     ...        ...
95         Joe Wright  4857       2   36588  Directing
96        Rob Minkoff  4858       2   18898  Directing
97       Lee Tamahori  4859       2    7256  Directing
98          Paul Feig  4860       2  116805  Directing
99       Peter Ramsey  4861       2  151007  Directing

[100 rows x 5 columns]

In [ ]:
directors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2349 entries, 0 to 2348
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   name        2349 non-null   object
 1   id          2349 non-null   int64 
 2   gender      2349 non-null   int64 
 3   uid         2349 non-null   int64 
 4   department  2349 non-null   object
dtypes: int64(3), object(2)
memory usage: 91.9+ KB


## Question 3: Check how many movies are present in iMDB.

In [ ]:
cur.execute('SELECT COUNT(*) FROM movies')
total_mov = cur.fetchone()[0]
print("Total number of movies: ",total_mov)




Total number of movies:  4773


Note/ Analysis :
There are 4773 movies in the IMDb dataset


## List out Top 10 highest budget made movies

In [ ]:
cur.execute("SELECT original_title, budget FROM movies ORDER BY budget DESC LIMIT 10")
top10_budget_movies = cur.fetchall()


df = pd.DataFrame(top10_budget_movies, columns=['Original Title', 'Budget'])
df



Original Title     Budget
0  Pirates of the Caribbean: On Stranger Tides  380000000
1     Pirates of the Caribbean: At World's End  300000000
2                      Avengers: Age of Ultron  280000000
3                             Superman Returns  270000000
4                                  John Carter  260000000
5                                      Tangled  260000000
6                                 Spider-Man 3  258000000
7                              The Lone Ranger  255000000
8                        The Dark Knight Rises  250000000
9       Harry Potter and the Half-Blood Prince  250000000

Note/ Analysis:
The highest budget movie is " Pirates of the Caribbean: On Stranger Tides"

#Question 4: Find these 3 directors: James Cameron ; Luc Besson ; John Woo

In [ ]:
cur.execute("SELECT * FROM directors WHERE Name IN ('James Cameron', 'Luc Besson', 'John Woo' )")
directors_data = cur.fetchall()
df = pd.DataFrame(directors_data, columns=['Name', 'ID', 'Gender', 'UID', 'Department'])
df



Name    ID  Gender    UID Department
0  James Cameron  4762       2   2710  Directing
1       John Woo  4893       2  11401  Directing
2     Luc Besson  4949       2     59  Directing

#Question 5: Find all directors with name starting with Steven.

In [ ]:
cur.execute("SELECT * FROM directors WHERE Name LIKE 'Steven%' ")
directors_data = cur.fetchall()
directors_df = pd.DataFrame(directors_data, columns=['Name', 'ID', 'Gender', 'UID', 'Department'])
(directors_df)


Name    ID  Gender    UID Department
0    Steven Spielberg  4799       2    488  Directing
1   Steven Soderbergh  4909       2   1884  Directing
2        Steven Brill  5013       2  32593  Directing
3     Steven Zaillian  5117       2   2260  Directing
4        Steven Quale  5216       2  93214  Directing
5       Steven Seagal  5221       2  23880  Directing
6  Steven E. de Souza  5390       2   1726  Directing
7    Steven Shainberg  5803       2  67795  Directing
8    Steven R. Monroe  6713       2  88039  Directing

Analysis/ Note:
There are 9 directors name which starts with Steven.

#Question 6: Count female directors.

In [ ]:
cur.execute("SELECT COUNT(*) FROM directors WHERE gender='Female'")
female_directors_count = cur.fetchone()[0]
(female_directors_count)

0

Note/Analysis: There are 0 female directors in the dataset.


#Question 7: Find the name of the 10th first women directors?


In [ ]:
cur.execute('SELECT Name FROM directors WHERE gender = "Female" LIMIT 10')
female_directors = cur.fetchall()

df = pd.DataFrame(female_directors, columns=['Name'])
df





Empty DataFrame
Columns: [Name]
Index: []

Note/Analysis: There is no women director

#Question 8: What are the 3 most popular movies?

In [ ]:
cur.execute("SELECT original_title, popularity FROM movies ORDER BY popularity DESC LIMIT 3")
most_3_popular_movies = cur.fetchall()
df = pd.DataFrame(most_3_popular_movies, columns=['Original Title', 'Popularity'])
df




Original Title  Popularity
0        Minions         875
1   Interstellar         724
2       Deadpool         514

Note/Analysis: The 3 most popular movies are Minions, Intrestellar and Deadpool.

#Question 9: What are the 3 most bankable movies?

In [ ]:
cur.execute(" SELECT original_title, revenue FROM movies ORDER BY revenue DESC LIMIT 3")
most_bankable_movies = cur.fetchall()
df= pd.DataFrame(most_bankable_movies, columns=['Original Title', 'Revenue'])
df



Original Title     Revenue
0         Avatar  2787965087
1        Titanic  1845034188
2   The Avengers  1519557910

Note/Analysis: Avatar, Titanic and The Avenger are the 3 most bankable movies.

#Question 10: What is the most awarded average vote since the January 1st, 2000?


In [ ]:
cur.execute("SELECT original_title, vote_average, release_date FROM movies WHERE release_date >= '2000-01-01' ORDER BY vote_average DESC LIMIT 1")
a= cur.fetchone()
a

('Sardaarji', 9.5, '2015-06-26')

Note/Analysis: Sardarji is the most awarded average vote since the January 1st, 2000.



#Question 11: Which movie(s) were directed by Brenda Chapman?

In [ ]:
sub_query=("SELECT original_title FROM movies WHERE director_id IN (SELECT id FROM directors WHERE name = 'Brenda Chapman')")
cur.execute(sub_query)
brenda_movies= cur.fetchall()
df = pd.DataFrame(brenda_movies, columns=['Original Title'])
df

Original Title
0          Brave

Note/Analysis: There is only one movie of  Brenda Chapman which is Brave

#Question 12: Whose director made the most movies?

In [ ]:
sub_query=("SELECT directors.name, COUNT(movies.id) as movie_count FROM movies JOIN directors ON movies.director_id = directors.id GROUP BY directors.name ORDER BY movie_count DESC LIMIT 1")
cur.execute(sub_query)
director= cur.fetchall()


df = pd.DataFrame(director, columns=['Director Name', 'Number of Movies'])
df


Director Name  Number of Movies
0  Steven Spielberg                27

Note/Analysis: Steven Spielberg is director who made the most movies which is 27.

#Question 13: Whose director is the most bankable?

In [ ]:
query=("SELECT directors.name, SUM(movies.revenue) as total_revenue FROM movies JOIN directors ON movies.director_id = directors.id GROUP BY directors.name ORDER BY total_revenue DESC LIMIT 1")
cur.execute(query)
director = cur.fetchall()



df = pd.DataFrame(director, columns=['Director Name', 'Total Revenue'])
df



Director Name  Total Revenue
0  Steven Spielberg     9147393164

Note/Analysis: Steven Spielberg is the most most bankable director.

#Budget Analysis
##Question 14 - List out Top 10 highest budget made movies


In [ ]:
query = "SELECT original_title, budget FROM movies ORDER BY budget DESC LIMIT 10"
cur.execute(query)
top10= cur.fetchall()

df = pd.DataFrame(top10, columns=['Original Title', 'Budget'])
df



Original Title     Budget
0  Pirates of the Caribbean: On Stranger Tides  380000000
1     Pirates of the Caribbean: At World's End  300000000
2                      Avengers: Age of Ultron  280000000
3                             Superman Returns  270000000
4                                  John Carter  260000000
5                                      Tangled  260000000
6                                 Spider-Man 3  258000000
7                              The Lone Ranger  255000000
8                        The Dark Knight Rises  250000000
9       Harry Potter and the Half-Blood Prince  250000000

<ipython-input-32-c61f0c8f89b5>:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.mean()


Budget    276300000.0
dtype: float64

#Popularity Analysis
##Question 15- List out Top 10 popularity on movies


In [ ]:
query = "SELECT original_title, popularity FROM movies ORDER BY popularity DESC LIMIT 10"
cur.execute(query)
popular_movies = cur.fetchall()

df = pd.DataFrame(popular_movies, columns=['Original Title', 'Popularity'])
df


Original Title  Popularity
0                                            Minions         875
1                                       Interstellar         724
2                                           Deadpool         514
3                            Guardians of the Galaxy         481
4                                 Mad Max: Fury Road         434
5                                     Jurassic World         418
6  Pirates of the Caribbean: The Curse of the Bla...         271
7                     Dawn of the Planet of the Apes         243
8              The Hunger Games: Mockingjay - Part 1         206
9                                         Big Hero 6         203

##Note/Analysis: Minions	is the most popular movie.

#Revenue Analysis
##Question 16- What are the top 10 Revenue movies


In [ ]:
query = "SELECT original_title, revenue FROM movies ORDER BY revenue DESC LIMIT 10"
cur.execute(query)
revenue_movies = cur.fetchall()

df= pd.DataFrame(revenue_movies, columns=['Original Title', 'Revenue'])
df


Original Title     Revenue
0                      Avatar  2787965087
1                     Titanic  1845034188
2                The Avengers  1519557910
3              Jurassic World  1513528810
4                   Furious 7  1506249360
5     Avengers: Age of Ultron  1405403694
6                      Frozen  1274219009
7                  Iron Man 3  1215439994
8                     Minions  1156730962
9  Captain America: Civil War  1153304495

Note/Analysis: Avatar is the most revenue movie.

#Voting Analysis
##Question 17 - Display most popular movies with vote_average


In [ ]:
query = "SELECT original_title, vote_average FROM movies ORDER BY vote_average DESC LIMIT 10"
cur.execute(query)
top10= cur.fetchall()

df = pd.DataFrame(top10, columns=['Original Title', 'Vote Average'])
df


Original Title  Vote Average
0          Stiff Upper Lips          10.0
1     Dancer, Texas Pop. 81          10.0
2                 Sardaarji           9.5
3            One Man's Hero           9.3
4  The Shawshank Redemption           8.5
5        There Goes My Baby           8.5
6     The Prisoner of Zenda           8.4
7             The Godfather           8.4
8                Fight Club           8.3
9          Schindler's List           8.3

Note/Analysis:
Stiff Upper Lips	ismost popular movies with vote
 average.

#Voting Analysis
##Question 18 - List out Particular movie for Voting average and votecount

In [ ]:
query = "SELECT original_title, vote_average, vote_count FROM movies WHERE original_title = 'Iron Man 3'"
cur.execute(query)
Iron_Man_3 = cur.fetchall()
df = pd.DataFrame(Iron_Man_3, columns=['Original Title', 'Vote Average', 'Vote Count'])
df


Original Title  Vote Average  Vote Count
0     Iron Man 3           6.8        8806

#Directors Analysis
##Question- 19 List out the director names with number of movies and revenue


In [ ]:
query = "SELECT directors.name, COUNT(movies.id) as movie_count, SUM(movies.revenue) as total_revenue FROM movies JOIN directors ON movies.director_id = directors.id GROUP BY directors.name ORDER BY total_revenue DESC"
cur.execute(query)
directors_details = cur.fetchall()

df= pd.DataFrame(directors_details, columns=['Director Name', 'Number of Movies', 'Total Revenue'])
df.head()


Director Name  Number of Movies  Total Revenue
0   Steven Spielberg                27     9147393164
1      Peter Jackson                 9     6498642820
2      James Cameron                 7     5883569439
3        Michael Bay                12     5832524638
4  Christopher Nolan                 8     4227483234

Note analysis: The max no. of movies by Steven Spielberg	which is 27.